From 2bc31d4d0f90fb8f31732d9e6148f2a1bec906b7 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:42:27 -0400
Subject: [ethernet] cxgb4: Allow >10G ports to have multiple queues

Message-id: <1410475447-94392-9-git-send-email-svemuri@redhat.com>
Patchwork-id: 93805
O-Subject: [RHEL7.1 PATCH BZ 1124947 008/108] cxgb4: Allow >10G ports to have multiple queues
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Based on original work by Divy Le Ray.

Signed-off-by: Kumar Sanghvi <kumaras@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 57d8b7649d119e8ad6a976a196a21cb4e5e2714c)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
index 7b626ec..475a1e6 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
@@ -5614,9 +5614,10 @@ static const struct pci_error_handlers cxgb4_eeh = {
  .resume         = eeh_resume,
 };
 
-static inline bool is_10g_port(const struct link_config *lc)
+static inline bool is_x_10g_port(const struct link_config *lc)
 {
- return (lc->supported & FW_PORT_CAP_SPEED_10G) != 0;
+ return (lc->supported & FW_PORT_CAP_SPEED_10G) != 0 ||
+        (lc->supported & FW_PORT_CAP_SPEED_40G) != 0;
 }
 
 static inline void init_rspq(struct sge_rspq *q, u8 timer_idx, u8 pkt_cnt_idx,
@@ -5640,7 +5641,7 @@ static void cfg_queues(struct adapter *adap)
  int i, q10g = 0, n10g = 0, qidx = 0;
 
  for_each_port(adap, i)
-  n10g += is_10g_port(&adap2pinfo(adap, i)->link_cfg);
+  n10g += is_x_10g_port(&adap2pinfo(adap, i)->link_cfg);
 
  /*
   * We default to 1 queue per non-10G port and up to # of cores queues
@@ -5655,7 +5656,7 @@ static void cfg_queues(struct adapter *adap)
   struct port_info *pi = adap2pinfo(adap, i);
 
   pi->first_qset = qidx;
-  pi->nqsets = is_10g_port(&pi->link_cfg) ? q10g : 1;
+  pi->nqsets = is_x_10g_port(&pi->link_cfg) ? q10g : 1;
   qidx += pi->nqsets;
  }
 
-- 
1.7.1